src/baldr/edgeinfo.cc

```diff
@@ -195,5 +208,15 @@ json::MapPtr EdgeInfo::json() const {
   return edge_info;
 }
 
+const std::unordered_map<TaggedName, std::string>& EdgeInfo::GetTags() const {
+  if (tagCache_.empty()) {
+    auto tags = GetTaggedNamesAndTypes();
+    for (const auto& tag : tags) {
+      tagCache_.emplace(static_cast<TaggedName>(tag.second), tag.first);
+    }
+  }
+  return tagCache_;
+}
```

Comment:
i like this approach much much better than what was done above in `GetTaggedNamesAndTypes`. i think we should move this caching into that function and indeed return a map of `Tag -> Value` like this does rather than `Value -> Tag` like the above method does. i know it will cause some work downstream but its clearly the right way to go. maybe as a first step just do the caching up there and then as a second step, in one isolated commit flip the key and the value in the returned unordered_map. what do you think?

Reply:
Well, I just realized we cannot use unordered_map here, because tags can have multiple values(tunnels is an example if I properly get their parsing code), but I think we can go with multimap here...

Reply:
yep exactly

Reply:
@SiarheiFedartsou  Hi, I am trying to understand this PR around z-level and why we cannot use unordered_map - could you help explain why "tags can have multiple values"? I couldn't find any such examples (Tag:key=tunnel) that you have described.. thanks a lot!

Reply:
@liyinxiao I meant tunnels in Valhalla tiles may have multiple names and we store all these names under the same tag name. "Tags" here is not directly mirrored by tags from OSM - it is what was generated by Mjolnir from OSM data.

Reply:
@SiarheiFedartsou  Very helpful, thanks!